src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/4.sql

```diff
@@ -0,0 +1,2159 @@
+--
+-- PARAMETERS
+--     @baseResourceSurrogateId
+--         * A bigint to which a value between [0, 80000) is added, forming a unique ResourceSurrogateId.
+--         * This value should be the current UTC datetime, truncated to millisecond precision, with its 100ns ticks component bitshifted left by 3.
+--     @resourceTypeId
+--         * The ID of the resource type (See ResourceType table)
+--     @resourceid
+--         * The resource ID (must be the same as the in the resource itself)
+--     @allowCreate
+--         * If false, an error is thrown if the resource does not already exist
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted
+--     @updatedDateTime
+--         * The last modified time in the resource
+--     @keepHistory
+--         * Whether the existing version of the resource should be preserved
+--     @requestMethod
+--         * The HTTP method/verb used for the request
+--     @rawResource
+--         * A compressed UTF16-encoded JSON document
+--     @isRawResourceMetaSet
+--         * Whether the rawResource json contains updated values for the meta property
+--     @resourceWriteClaims
+--         * Claims on the principal that performed the write
+--     @compartmentAssignments
+--         * Compartments that the resource is part of
+--     @referenceSearchParams
+--         * Extracted reference search params
+--     @tokenSearchParams
+--         * Extracted token search params
+--     @tokenTextSearchParams
+--         * The text representation of extracted token search params
+--     @stringSearchParams
+--         * Extracted string search params
+--     @numberSearchParams
+--         * Extracted number search params
+--     @quantitySearchParams
+--         * Extracted quantity search params
+--     @uriSearchParams
+--         * Extracted URI search params
+--     @dateTimeSearchParms
+--         * Extracted datetime search params
+--     @referenceTokenCompositeSearchParams
+--         * Extracted reference$token search params
+--     @tokenTokenCompositeSearchParams
+--         * Extracted token$token tokensearch params
+--     @tokenDateTimeCompositeSearchParams
+--         * Extracted token$datetime search params
+--     @tokenQuantityCompositeSearchParams
+--         * Extracted token$quantity search params
+--     @tokenStringCompositeSearchParams
+--         * Extracted token$string search params
+--     @tokenNumberNumberCompositeSearchParams
+--         * Extracted token$number$number search params
+--
+-- RETURN VALUE
+--         The version of the resource as a result set. Will be empty if no insertion was done.
+--
+CREATE PROCEDURE dbo.UpsertResource
+    @baseResourceSurrogateId bigint,
+    @resourceTypeId smallint,
+    @resourceId varchar(64),
+    @eTag int = NULL,
+    @allowCreate bit,
+    @isDeleted bit,
+    @keepHistory bit,
+    @requestMethod varchar(10),
+    @rawResource varbinary(max),
+    @isRawResourceMetaSet bit = 0,
```

Comment:
Could we skip this parameter? I think we can set the colum in the databse based solely on whether the final version is 1.

Reply:
If we skip it, can skip all the duplicate TvpRowRegenerators...

Reply:
The GetResource call is a stored proc, so I don't think we can get rid of those quite yet.

Reply:
If we do (2) [here](https://github.com/microsoft/fhir-server/pull/1178#discussion_r485111502) then I don't think the codegen for this stored procedure is affected.

Reply:
Actually, we have a similar problem here. If an "old" FHIR server calls this new stored procedure, then we will incorrectly assume that the version is set in the resource.

Reply:
Sorry for thinking out loud here. We actually require that all FHIR servers be upgraded to the new version _before_ we begin the schema upgrade, so this may not be an issue.